src/compiler/parser.rs

```diff
@@ -1171,19 +1194,67 @@ impl ParserType {
         self.parse_postfix_type_or_higher()
     }
 
-    fn parse_union_or_intersection_type(
+    fn parse_function_or_constructor_type_to_error(
+        &self,
+        is_in_union_type: bool,
+    ) -> Option<TypeNode> {
+        None
+    }
+
+    fn parse_union_or_intersection_type<TReturn: Into<TypeNode>>(
         &mut self,
+        operator: SyntaxKind, /*SyntaxKind.BarToken | SyntaxKind.AmpersandToken*/
         parse_constituent_type: fn(&mut ParserType) -> TypeNode,
+        create_type_node: fn(&NodeFactory, &ParserType, NodeArray) -> TReturn,
     ) -> TypeNode {
-        parse_constituent_type(self)
+        let pos = self.get_node_pos();
+        let is_union_type = operator == SyntaxKind::BarToken;
+        let has_leading_operator = self.parse_optional(operator);
+        let mut type_: Option<TypeNode> = Some(if has_leading_operator {
+            self.parse_function_or_constructor_type_to_error(is_union_type)
+                .unwrap_or_else(|| parse_constituent_type(self))
+        } else {
+            parse_constituent_type(self)
+        });
+        if self.token() == operator || has_leading_operator {
+            let mut types: Vec<Node> = vec![type_.take().unwrap().into()];
```

Comment:
Here I gave `type_` an `Option` wrapper so that I could move it here (by "taking" it) but still reassign to it later (to be able to return it at the end of the function)